src/mscorlib/src/System/Collections/Generic/List.cs

```diff
@@ -1031,6 +1028,45 @@ public void Sort(IComparer<T> comparer)
             return true;
         }
 
+        private void AddEnumerable(IEnumerable<T> enumerable)
+        {
+            Contract.Assert(enumerable != null);
+            Contract.Assert(!(enumerable is ICollection<T>), "We should have optimized for this beforehand.");
+
+            using (IEnumerator<T> en = enumerable.GetEnumerator())
+            {
+                T[] items = _items;
+                int size = _size;
+
+                // Read in items from the enumerator, only updating fields
+                // when we run out of space.
+
+                try
+                {
+                    while (en.MoveNext())
+                    {
+                        if (size == items.Length)
+                        {
+                            _size = size;
+                            EnsureCapacity(size + 1);
+                            items = _items;
+                        }
+
+                        // Note: It's important we increment size after Current is called.
+                        // If that throws an exception we don't want to do the increment.
+                        items[size] = en.Current;
+                        size++;
+                    }
+                }
+                finally
+                {
+                    // Make a final update to _size and _version after we've finished adding.
+                    _size = size;
+                    _version++;
```

Comment:
> EnsureCapacity always updates _items, and we always update the local variable via items = _items after that

But those lines aren't relevant to the case I'm talking about.  If the call to MoveNext increases the capacity of the list, items will now be different from _items, we won't enter that if block (as we will already have enough capacity since it was just resized), and we'll store the items into the wrong array.